File: rxjava/src/main/java/xyz/truenight/databinding/rxjava/RxRefresh.java
package xyz.truenight.databinding.rxjava;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.ObservableTransformer;
import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Predicate;
import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.Subject;
import xyz.truenight.utils.Utils;

/**
 * Created by true
 * date: 17/11/2017
 * time: 20:19
 * <p>
 * Copyright © Mikhail Frolov
 */

public class RxRefresh implements Action {

    private final Subject<Boolean> refresh = PublishSubject.<Boolean>create().toSerialized();

    private Consumer<Throwable> handler;

    private RxRefresh() {

    }

    public static RxRefresh create() {
        return new RxRefresh();
    }

    public static <T> ObservableTransformer<T, T> merge(RxRefresh refresh, RxRefresh... toMerge) {
        List<Subject<Boolean>> merged = new ArrayList<>();
        merged.add(refresh.refresh);
        if (Utils.isNotEmpty(toMerge)) {
            for (RxRefresh rxRefresh : toMerge) {
                merged.add(rxRefresh.refresh);
            }
        }
        Observable<Boolean> merge = Observable.merge(merged);
        return observable -> {
            // handler will be taken from primary RxRefresh
            if (refresh.handler != null) {
                observable = observable.doOnError(refresh.handler);
            }
            return observable
                    .retryWhen(o -> o.flatMap(t -> merge.firstElement().toObservable()))
                    .repeatWhen(o -> o.flatMap(t -> merge.firstElement().toObservable()));
        };
    }

    public RxRefresh errorHandler(Consumer<Throwable> handler) {
        this.handler = handler;
        return this;
    }

    public <T> ObservableTransformer<T, T> transformer() {
        return observable -> {
            if (handler != null) {
                observable = observable.doOnError(handler);
            }
            return observable
                    .retryWhen(o -> o.flatMap(t -> refresh.firstElement().toObservable()))
                    .repeatWhen(o -> o.flatMap(t -> refresh.firstElement().toObservable()));
        };
    }

    public <T> ObservableTransformer<T, T> mergedTransformer(RxRefresh... toMerge) {
        return merge(this, toMerge);
    }

    public <T> ObservableTransformer<T, T> transformerWithCondition(Predicate<Boolean> predicate) {
        return observable -> {
            if (handler != null) {
                observable = observable.doOnError(handler);
            }
            return observable
                    .retryWhen(o -> o.flatMap(t -> refresh
                            .takeWhile(predicate).firstElement().toObservable()))
                    .repeatWhen(o -> o.flatMap(t -> refresh.takeWhile(predicate).firstElement().toObservable()));
        };
    }

    @Override
    public void run() {
        refresh.onNext(false);
    }

    public Action withDelay(long delay, TimeUnit timeUnit) {
        return () -> Observable.timer(delay, timeUnit).doFinally(this).subscribe();
    }
}

File: rxjava/src/main/java/xyz/truenight/databinding/rxjava/RxRefresh.kt
package xyz.truenight.databinding.rxjava

import io.reactivex.*
import io.reactivex.functions.Action
import io.reactivex.functions.Consumer
import io.reactivex.functions.Predicate
import io.reactivex.processors.PublishProcessor
import java.util.concurrent.TimeUnit

/**
 * Created by true
 * date: 17/11/2017
 * time: 20:19
 *
 *
 * Copyright © Mikhail Frolov
 */

class RxRefresh private constructor() : Action {

    internal val refresh = PublishProcessor.create<Boolean>().toSerialized()

    internal var handler: Consumer<Throwable>? = null

    fun errorHandler(handler: Consumer<Throwable>): RxRefresh {
        this.handler = handler
        return this
    }

    fun <T> transformer(): ObservableTransformer<T, T> {
        return ObservableTransformer { observable ->
            observable.doOnError { handler?.accept(it) }
                    .retryWhen { it.flatMap { refresh.firstElement().toObservable() } }
                    .repeatWhen { it.flatMap { refresh.firstElement().toObservable() } }
        }
    }

    fun <T> mergedTransformer(vararg toMerge: RxRefresh): ObservableTransformer<T, T> {
        return mergeWith(*toMerge)
    }

    fun <T> transformerWithCondition(predicate: Predicate<Boolean>): ObservableTransformer<T, T> {
        return ObservableTransformer { observable ->
            observable.doOnError { handler?.accept(it) }
                    .retryWhen { it.flatMap { refresh.filter(predicate).firstElement().toObservable() } }
                    .repeatWhen { it.flatMap { refresh.filter(predicate).firstElement().toObservable() } }
        }
    }

    override fun run() {
        refresh.onNext(false)
    }

    fun withDelay(delay: Long, timeUnit: TimeUnit): Action {
        return Action { Observable.timer(delay, timeUnit).doFinally(this).subscribe() }
    }

    companion object {

        fun create(): RxRefresh {
            return RxRefresh()
        }

        private fun <T> RxRefresh.mergeWith(vararg toMerge: RxRefresh): ObservableTransformer<T, T> {

            val merge = Flowable.merge(
                    mutableListOf<Flowable<Boolean>>().apply {
                        add(refresh)
                        toMerge.mapTo(this) { it.refresh }
                    })
            return ObservableTransformer { observable ->
                // handler will be taken from primary RxRefresh
                observable.doOnError { handler?.accept(it) }
                        .retryWhen { it.flatMap { merge.firstElement().toObservable() } }
                        .repeatWhen { it.flatMap { merge.firstElement().toObservable() } }
            }
        }
    }
}

fun <T> Single<T>.refresh(refresh: RxRefresh) =
        doOnError { refresh.handler?.accept(it) }
                .retryWhen { it.flatMap { refresh.refresh.firstElement().toFlowable() } }
                .repeatWhen { it.flatMap { refresh.refresh.firstElement().toFlowable() } }

fun <T> Flowable<T>.refresh(refresh: RxRefresh) =
        doOnError { refresh.handler?.accept(it) }
                .retryWhen { it.flatMap { refresh.refresh.firstElement().toFlowable() } }
                .repeatWhen { it.flatMap { refresh.refresh.firstElement().toFlowable() } }

fun <T> Observable<T>.refresh(refresh: RxRefresh) =
        doOnError { refresh.handler?.accept(it) }
                .retryWhen { it.flatMap { refresh.refresh.firstElement().toObservable() } }
                .repeatWhen { it.flatMap { refresh.refresh.firstElement().toObservable() } }

fun Completable.refresh(refresh: RxRefresh) =
        doOnError { refresh.handler?.accept(it) }
                .retryWhen { it.flatMap { refresh.refresh.firstElement().toFlowable() } }
                .repeatWhen { it.flatMap { refresh.refresh.firstElement().toFlowable() } }
